doc/developer/design/20220404_metadata_storage.md

```diff
@@ -0,0 +1,304 @@
+# Metadata database for Materialize Platform
+
+## Summary
+
+Each [Materialize Platform layer](../platform/architecture-db.md) in each
+customer environment in Materialize Platform needs to durably record its
+metadata. What database to use is an open question. The correctness,
+performance, scalability, reliability, and availability goals of Materialize
+Platform heavily constrain our options. This document details those constraints
+and proposes a candidate system for further evaluation.
+
+## Constraints
+
+### Hard requirements
+
+The following are the non-negotiable requirements that flow from Materialize
+Platform's product goals.
+
+* **Resilience to single availability zone outages.** Materialize Platform must
+  be resilient to an outage in a single availability zone of a cloud provider
+  region. Therefore the metadata database must be resilient to outages in a
+  single availability zone.
+
+* **Strong consistency.** Materialize Platform intends to offer strict
```

Comment:
I think the premises of this design doc are:

1. The current SQLite-on-EBS metadata storage solution does not meet our availability goals.
2. Manual failover is not safe enough to provide the consistency guarantees we need to pass the Jepsen tests.

Let's definitely talk about challenging premise (1). How bad would it be to keep rolling with SQLite on EBS for the first version of Materialize Platform?

I'd love for you to prove me wrong on (2), but I'm pretty sure it's fundamentally impossible to make manual failover truly safe in a cloud environment.?

Reply:
Let's expand on (1) more here, and I'd love to benefit from @alex-hunt-materialize 's experience as well: are our availability concerns about EBS strictly tied to the AZ limitation? My operational experience with EBS as a _reliable_ store has been limited: I have experienced durability issues with EBS on the AZ front, but also availability issues allocating EBS volumes and spiky performance.
  - If our concerns _are_ strictly tied to AZs, is EFS an option? It may be disqualified for other reasons (I have experienced wildly variable performance on EFS, for instance, but that was a few years ago), but it should be something to consider if durability is our primary concern here.

I agree on (2) that manual failover would not do the job as far as passing a Jepsen test: we could certainly bodge it through well enough for likely customer scenarios, but to honestly make the kinds of claims Jepsen tests I don't think we can manage it manually.

Reply:
> I agree on (2) that manual failover would not do the job as far as passing a Jepsen test: we could certainly bodge it through well enough for likely customer scenarios, but to honestly make the kinds of claims Jepsen tests I don't think we can manage it manually.

Ok, great!

> If our concerns are strictly tied to AZs, is EFS an option? It may be disqualified for other reasons (I have experienced wildly variable performance on EFS, for instance, but that was a few years ago), but it should be something to consider if durability is our primary concern here.

I think it is an option! Building correct systems on top of NFS scares me deeply, but as far as I can tell running SQLite on NFS is generally thought to be safe. 🤷🏽‍♂️

Reply:
NFS scares me too. I've used it on-prem and in AWS' EFS implementation, and it's always extremely broken. Hung mounts are not something I ever want to deal with in Kubernetes.

Please, no NFS. I'd much rather be stuck with all the problems of local EBS than NFS.

Reply:
EBS is less reliable than most other AWS storage mechanisms (EFS is much worse), but the main reasons to avoid it are operational rather than reliability. It's just super hard to maintain and modify, especially in K8S.